aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -599,6 +599,98 @@ Tensor& diff_out(const Tensor& self, int64_t n, int64_t dim, const c10::optional
+    int64_t direction = maybe_wrap_dim(axis[i], self.dim());
+
+    if (spacing[i].dim()==1) {
+      std::vector<int64_t> shape(self.dim(),1);
+      shape[ direction ] = -1;
+      auto ax_dx = spacing[i].diff(1,0);
+      auto dx1 = at::slice(ax_dx, 0, 0, -1);
+      auto dx2 = at::slice(ax_dx, 0, 1);
+      auto a = (   -dx2    / (dx1*(dx1+dx2)) ).reshape(shape);
+      auto b = ( (dx2-dx1) / (dx1*dx2)       ).reshape(shape);
+      auto c = (    dx1    / (dx2*(dx1+dx2)) ).reshape(shape);
+
+      auto center  = a*at::slice(self, direction, 0, -2) + b*at::slice(self, direction , 1, -1) + c*at::slice(self, direction ,2);
+      auto prepend = (at::slice(self, direction, 1, 2  ) - at::slice(self, direction, 0, 1   )) / ax_dx[0]  ;
+      auto append  = (at::slice(self, direction, -1    ) - at::slice(self, direction, -2, -1 )) / ax_dx[-1] ;
+      result.push_back(prepend_append_on_dim(center, prepend, append, direction));
+    } else {
+      auto ax_dx = spacing[i];
+      auto center  = (at::slice(self,direction, 2   ) - at::slice(self, direction, 0, -2 ) ) / (2.0 *ax_dx);
+      auto prepend = (at::slice(self,direction, 1, 2) - at::slice(self, direction, 0, 1  ) ) / ax_dx  ;
+      auto append  = (at::slice(self,direction, -1  ) - at::slice(self, direction, -2, -1) ) / ax_dx ;
+      result.push_back(prepend_append_on_dim(center, prepend, append, direction));
+    }
+  }
+  return result;
+}
+
+std::vector<Tensor> gradient(const Tensor& self,
+                             TensorList spacing,
+                             c10::optional<int64_t> dim,
+                             int64_t edge_order) {
+
+  std::vector<int64_t> axis;
+  if (!dim.has_value()) {
+    for (const auto i : c10::irange(spacing.size())) {
+      axis.push_back(i);
+    }
+    return self.gradient(spacing, IntArrayRef(axis), edge_order);
+  }
+  axis.push_back(dim.value());
+  return self.gradient(spacing, IntArrayRef(axis), edge_order);
+}
+
+std::vector<Tensor> gradient(const Tensor& self,
+                             double unit_size,
+                             IntArrayRef axis,
+                             int64_t edge_order) {
+  std::vector<Tensor> spacing(axis.size());
+  for (const auto i : c10::irange(spacing.size())) {
+    Tensor tensor_for_unit_size = at::zeros(1, self.options().dtype(at::kFloat));
+    tensor_for_unit_size[0] = unit_size;
+    spacing[i] = tensor_for_unit_size[0];
+  }
+  return self.gradient(spacing, axis, edge_order);
+}
+
+std::vector<Tensor> gradient(const Tensor& self,
+                             c10::optional<double> unit_size,
+                             c10::optional<int64_t> dim,
+                             int64_t edge_order) {
+  double unit_size_value = !unit_size.has_value() ? 1.0 : unit_size.value();
+  std::vector<Tensor> spacing(dim.has_value() ? 1 : self.dim() );
+  for (const auto i : c10::irange(spacing.size())) {
```

Comment:
Like the above: how could a list of scalars be handled without creating tensors for each scalar value?